### ISSUE TITLE
Bug: implicit any flags renamed destructuring from async/await within loops when shadowing variable names

### ISSUE THREAD

**Issue by admwx7** [microsoft/TypeScript#36666]
**NOTE**: this issue is quite similar to https://github.com/microsoft/TypeScript/issues/32278 though I could only reproduce my exact issue when using a loop and async functions with a let/var.  The conditions were different enough I though it made sense to open a separate issue.

**TypeScript Version:**  3.8.0-dev.20200205

**Search Terms:**
circular, implicit, referenced directly

**Code**

```ts
async function inner(param1: string | undefined): Promise<{ param1?: string }> {
  return { param1: '' };
}

async function outer() {
  let param1;
  do {
    const { param1: newParam } = await inner(param1);
    param1 = newParam;
  } while (param1)
}
```

**Expected behavior:**
No error

**Actual behavior:**
`newParam` is flagged with the following error:
```
const newParam: any
'newParam' implicitly has type 'any' because it does not have a type annotation and is referenced directly or indirectly in its own initializer.
```

**Playground Link:**
[playground](http://www.typescriptlang.org/play/?ts=3.8.0-dev.20200205&ssl=1&ssc=1&pln=11&pc=2#code/IYZwngdgxgBAZgV2gFwJYHsI1RCBTAJwAoAHYA4AWwEYAuGEZAnAcxgB8YkATPOHPNwCU9AAoF0lVCDwAeAN4wyFGgH56jZhDYBfAHwx5AKBgwCeZAgJZFyqnRgByRzB0BuIzqNHQkWIhQMLHQEZEIiIUMTGAAbCyVyew9TbnQo01MoTEZDBJUHfAB3UUTKVxgAXhhgQuBUZGxccLsaIWSMlupKmCKSlXadGEKAC1Q4mFJS6iFPIA)

**Related Issues:** 
https://github.com/microsoft/TypeScript/issues/33191
https://github.com/microsoft/TypeScript/issues/29414
https://github.com/microsoft/TypeScript/issues/32278

**Comment by RyanCavanaugh** [comment]
This is definitely a circularity, thus the error. We can avoid *some* circularity errors but not all of them, but this isn't one we're equipped to avoid.
